src/core/qgsdataprovider.h

```diff
@@ -544,6 +569,18 @@ class CORE_EXPORT QgsDataProvider : public QObject
      */
     QString mDataSourceURI;
 
+    /**
+     * Contains collected Metadata for the Provider source
+     * \brief A structured metadata store for a map layer.
+     * \note
+     *  - QgsVectorLayer::setDataProvider
+     *  - QgsRasterLayer::setDataProvider
+     * \see layerMetadata
+     * \see setLayerMetadata
+    * \since QGIS 3.0
+    */
+    QgsLayerMetadata mLayerMetadata;
```

Comment:
I'd drop this, and just return a default constructed QgsLayerMetadata in the base class (and have the base class  setLayerMetadata do nothing), and leave it all up to individual provider subclasses how they want to handle this. 

I'm also thinking we should add a companion QgsVectorDataProvider::Capability flag for "GetMetadata" and "SetMetadata" to indicate whether individual providers can take advantage of these methods.

Reply:
But the need is that the provider should fill this with something useful, so that the Layer can retrieve it.

Reply:
The thing is - most providers won't do anything with this, and the metadata will be empty. I think it's up to individual providers to store the metadata member variable and implement an override of the getter/setter, if it's appropriate for that provider. I think only the virtual methods belong in the base class and not the implementation detail (the member variable)

Reply:
But how should the Layer, in setDataProvider, retrieve it?
That is here the problem that needs to be resolved.

Reply:
That's up to the provider. E.g. for spatialite provider you would add a `QgsLayerMetadata mMetadata` private member to the provider, which is populated at the same time the provider determines the layer's fields, crs, etc. Then the spatialite provider would implement a layerMetadata() getter which returns mMetadata. Other providers may not need the member variable, and e.g. may dynamically populate a new QgsLayerMetadata object only when their metadata() override is called.  

Presumably somewhere in your current work you're calling setLayerMetadata() within the spatialite provider - you'd replace that with just setting the spatialite provider mMetadata member variable which is returned by the layerMetadata() override.

Reply:
In other words in QgsDataProvider:
remove:     QgsLayerMetadata mLayerMetadata;
change to:
    virtual const QgsLayerMetadata &layerMetadata() const { return QgsLayerMetadata(); };
    virtual void setLayerMetadata( const QgsLayerMetadata &layerMetadata ) const = 0;

And in a Provider:
    const QgsLayerMetadata &layerMetadata() const override { return mLayerMetadata; };
    void setLayerMetadata( const QgsLayerMetadata &layerMetadata )  const override { mLayerMetadata = layerMetadata; };

correct?

Reply:
Yep, with the one exception that I think setLayerMetadata shouldn't be pure virtual. I.e.

    virtual void setLayerMetadata( const QgsLayerMetadata &layerMetadata ) { Q_UNUSED( layerMetadata ); }

since most providers won't be able to do anything with this. That's also why I think we need a new capability flag to indicate if a provider has support for storing metadata.

I'd also drop the const here and return a bool (success or not)

Reply:
In other word this:

virtual bool setLayerMetadata( const QgsLayerMetadata &layerMetadata ) { Q_UNUSED( layerMetadata ); return false; }

Reply:
Perfect!